src/test/java/uk/gov/pay/products/resources/ProductResourceTest.java

```diff
@@ -0,0 +1,95 @@
+package uk.gov.pay.products.resources;
+
+import org.jetbrains.annotations.NotNull;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import uk.gov.pay.products.client.publicauth.TokenResponse;
+import uk.gov.pay.products.config.ProductsConfiguration;
+import uk.gov.pay.products.fixtures.ProductEntityFixture;
+import uk.gov.pay.products.model.Product;
+import uk.gov.pay.products.service.ProductFactory;
+import uk.gov.pay.products.service.ProductFinder;
+import uk.gov.pay.products.util.ProductType;
+import uk.gov.pay.products.validations.ProductRequestValidator;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation.Builder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static uk.gov.pay.products.util.RandomIdGenerator.randomUuid;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ProductResourceTest {
+
+    private static final String PUBLICAUTH_URL = "https://publicauth.url";
+    private ProductResource productResource;
+    
+    @Mock
+    private ProductRequestValidator productRequestValidator;
+
+    @Mock
+    private ProductFactory productFactory;
+    
+    @Mock
+    private Client client;
+    
+    @Mock
+    private ProductsConfiguration productsConfiguration;
+
+    @Before
+    public void setUp() {
+        when(productsConfiguration.getEmailAddress()).thenReturn("pay-products@gov.uk");
+        when(productsConfiguration.getPublicAuthUrl()).thenReturn(PUBLICAUTH_URL);
+
+        productResource = new ProductResource(productRequestValidator, productFactory, client, productsConfiguration);
+    }
+    
+    @Test
+    public void shouldReturn500_whenFailingToUpdateApiKeyForAPaymentLink() {
```

Comment:
Feels a bit odd that we have a unit test just for this one thing but everything else is tested by integration tests.

Reply:
This is a special case where a product is deleted from the database just before Products application replaces an old API token with a new API token for the product. It is difficult to do this in integration test. That's why I am able to use mock in unit test to create this kind of scenario.